db/snapshot-ctl.cc

```diff
@@ -144,82 +147,57 @@ future<> snapshot_ctl::clear_snapshot(sstring tag, std::vector<sstring> keyspace
-                    continue;
-                }
-
-                auto& rp = _result.at(vp.first);
-                for (auto&& cf: vp.second) {
-                    if (auto [ignored, added] = rp.try_emplace(cf.first, std::move(cf.second)); added) {
-                        continue;
-                    }
-                    auto& rcf = rp.at(cf.first);
-                    rcf.live = cf.second.live;
-                    rcf.total = cf.second.total;
-                }
+        future<> operator()(const std::vector<database::snapshot_details_result>& value) {
+            for (auto&& vp : value) {
+                _result.emplace(std::move(vp));
             }
             return make_ready_future<>();
         }
-        snapshot_map get() && {
+
+        snapshot_acc get() && {
             return std::move(_result);
         }
     };
 
-    return run_snapshot_list_operation([this] {
-        return _db.map_reduce(snapshot_reducer(), [] (database& db) {
-            auto local_snapshots = make_lw_shared<snapshot_map>();
-            return parallel_for_each(db.get_column_families(), [local_snapshots] (auto& cf_pair) {
-                return cf_pair.second->get_snapshot_details().then([uuid = cf_pair.first, local_snapshots] (auto map) {
-                    for (auto&& snap_map: map) {
-                        auto [it, ignored] = local_snapshots->try_emplace(snap_map.first);
-                        it->second.emplace(uuid, snap_map.second);
-                    }
-                    return make_ready_future<>();
-                });
-            }).then([local_snapshots] {
-                return make_ready_future<snapshot_map>(std::move(*local_snapshots));
-            });
-        }).then([this] (snapshot_map&& map) {
-            std::unordered_map<sstring, std::vector<snapshot_ctl::snapshot_details>> result;
-            for (auto&& pair: map) {
-                std::vector<snapshot_ctl::snapshot_details> details;
-
-                for (auto&& snap_map: pair.second) {
-                    auto& cf = _db.local().find_column_family(snap_map.first);
-                    details.push_back({ snap_map.second.live, snap_map.second.total, cf.schema()->cf_name(), cf.schema()->ks_name() });
-                }
-                result.emplace(pair.first, std::move(details));
-            }
-
-            return make_ready_future<std::unordered_map<sstring, std::vector<snapshot_ctl::snapshot_details>>>(std::move(result));
+    co_return co_await run_snapshot_list_operation([this] () -> future<snapshot_map> {
+        snapshot_acc collected = co_await _db.map_reduce(snapshot_reducer(), [] (database& db) -> future<std::vector<database::snapshot_details_result>> {
+            co_return co_await db.get_snapshot_details();
         });
+
+        snapshot_map result;
+        for (auto&& r : collected) {
+            result[r.snapshot_name].emplace_back(std::move(r.details));
```

Comment:
Does it make sense to add `co_await coroutine::maybe_yield();` here as well?

Reply:
Yeah, it does iterate over the whole set of snapshots